lib/hanami/cli.rb

```diff
@@ -45,27 +45,57 @@ def initialize(registry)
     #
     # @param arguments [Array<string>] the command line arguments (defaults to `ARGV`)
     # @param out [IO] the standard output (defaults to `$stdout`)
+    # @param err [IO, NilClass] the standard output or the standard error or nil (default)
+    # @param kernel [Module, NilClass] optional replacement for Kernel, on which #exit() is called.
     #
     # @since 0.1.0
-    def call(arguments: ARGV, out: $stdout)
+    # rubocop:disable Metrics/AbcSize
+    # rubocop:disable Metrics/MethodLength
+    def call(arguments: ARGV, out: $stdout, **opts)
+      err = opts[:err]
+      @kernel = opts[:kernel]
+
+      original_argv = ARGV.dup
+
       result = commands.get(arguments)
 
       if result.found?
-        command, args = parse(result, out)
-
-        result.before_callbacks.run(command, args)
-        command.call(args)
-        result.after_callbacks.run(command, args)
+        call_command(result, out)
       else
-        usage(result, out)
+        exit_code = 0
+        invalid_commands = original_argv.reject { |a| a.start_with?('-') }
+        if invalid_commands.size.positive?
+          err&.puts("Error:\n")
+          err&.puts("  Unknown command(s): #{invalid_commands.join(', ')}")
+          err&.puts
+          exit_code = 1
+        end
+        usage(result, out, exit_code)
```

Comment:
Thoughts on extracting this to a helper method? Might help us avoid disabling `Metrics/AbcSize` and `Metrics/MethodLength` too.